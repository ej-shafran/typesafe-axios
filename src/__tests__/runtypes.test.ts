import { taxios } from "..";

import { String, Record, Array, Union, Literal, Static } from "runtypes";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = Record({
  id: String,
  username: String,
  email: String,
  address: Record({
    street: String,
    zipcode: String.optional(),
  }),
  followers: Array(
    Record({
      id: String,
      role: Union(Literal("admin"), Literal("guest")).optional(),
    })
  ),
});

type TYPE = Static<typeof SCHEMA>;
function ensure(_: TYPE) { }

describe("Runtypes interop", () => {
  it("passes along data with the correct type", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);

    const response = await taxios(SCHEMA).get(UTILS.API);

    expect(response.data.success).toBe(true);
    if (response.data.success) {
      expect(response.data.value).toEqual(UTILS.VALID);
      ensure(response.data.value);
    }
  });

  it("throws a validation error when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    const response = await taxios(SCHEMA).get(UTILS.API);

    expect(response.data.success).toBe(false);
    if (!response.data.success) {
      expect(response.data.message).toMatchSnapshot();
    }
  });
});
