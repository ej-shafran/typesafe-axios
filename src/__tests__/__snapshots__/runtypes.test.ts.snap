// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runtypes interop throws a validation error when the data is invalid 1`] = `
"Validation failed:
{
  "email": "Expected string, but was missing",
  "address": "Expected { street: string; zipcode?: string; }, but was string"
}.
Object should match { id: string; username: string; email: string; address: { street: string; zipcode?: string; }; followers: { id: string; role?: "admin" | "guest"; }[]; }"
`;
